README.md
[![Travis-CI Build Status](https://travis-ci.org/ncss-tech/soilDB.svg?branch=master)](https://travis-ci.org/ncss-tech/soilDB)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/soilDB)](https://cran.r-project.org/package=soilDB)
[![Total_Downloads](http://cranlogs.r-pkg.org/badges/grand-total/soilDB)](https://cran.r-project.org/package=soilDB)

## Installation

Get the stable version from CRAN:

`install.packages('soilDB', dep=TRUE)`

Get the development version from Github:

`remotes::install_github("ncss-tech/soilDB", dependencies=FALSE, upgrade=FALSE, build=FALSE)`

## Website
http://ncss-tech.github.io/AQP/


## Important changes in soilDB-2.0

### NASIS

 * low-level functions return empty `data.frame` objects when local database (or selected set) is empty
 * fetchNASIS() is now a wrapper around pedon and component "fetch" functions
 * uncode() is now used in all queries to local NASIS database
 * many (coded) column names in pedon @site and @horizons have changed
 * lower cases is used for all un-coded data
 

## Database APIs
    
  * SDA
    + [`SDA_query`](http://ncss-tech.github.io/soilDB/docs/reference/SDA_query.html)
    + ['SDA_query_features`](http://ncss-tech.github.io/soilDB/docs/reference/SDA_query_features.html)

  * SSURGO/KSSL via SoilWeb
    + [`fetchKSSL`](http://ncss-tech.github.io/soilDB/docs/reference/fetchKSSL.html)
    + [`fetchOSD`](http://ncss-tech.github.io/soilDB/docs/reference/fetchOSD.html)
    + [`siblings`](http://ncss-tech.github.io/soilDB/docs/reference/siblings.html)
    + [`OSDquery`](http://ncss-tech.github.io/soilDB/docs/reference/OSDquery.html) 
    + [`seriesExtent`](http://ncss-tech.github.io/soilDB/docs/reference/seriesExtent.html)
    
  * SCAN/SNOTEL
    + [`fetchSCAN`](http://ncss-tech.github.io/soilDB/docs/reference/fetchSCAN.html)
    + [`SCAN_SNOTEL_metadata`](http://ncss-tech.github.io/soilDB/docs/reference/SCAN_SNOTEL_metadata.html)
    
  * Henry Mount Soil and Water Database
    + [`fetchHenry`](http://ncss-tech.github.io/soilDB/docs/reference/fetchHenry.html)
    
  * NASIS local database
    + [`fetchNASIS`](http://ncss-tech.github.io/soilDB/docs/reference/fetchNASIS.html)
    
  * NASIS WWW interface
    + [`parseWebReport`](http://ncss-tech.github.io/soilDB/docs/reference/parseWebReport.html)
    + [`fetchNASISWebReport`](http://ncss-tech.github.io/soilDB/docs/reference/fetchLIMS_component.html)


## Utility Functions

  * [`estimateSTR`](http://ncss-tech.github.io/soilDB/docs/reference/estimateSTR.html)
  * [`STRplot`](http://ncss-tech.github.io/soilDB/docs/reference/STRplot.html)
  * [`KSSL_VG_model`](http://ncss-tech.github.io/soilDB/docs/reference/KSSL_VG_model.html)
  * [`simplfyFragmentData`](http://ncss-tech.github.io/soilDB/docs/reference/simplfyFragmentData.html)
  * [`simplifyColorData`](http://ncss-tech.github.io/soilDB/docs/reference/simplifyColorData.html)
  * [`uncode`](http://ncss-tech.github.io/soilDB/docs/reference/uncode.html)
  * [`code`](http://ncss-tech.github.io/soilDB/docs/reference/uncode.html)


## Related Documentation
 * [fetchKSSL](http://ncss-tech.github.io/AQP/soilDB/KSSL-demo.html)
 * [SDA_query](http://ncss-tech.github.io/AQP/soilDB/SDA-tutorial.html)
 * [fetchOSD](http://ncss-tech.github.io/AQP/sharpshootR/OSD-dendrogram.html)
 * [SCAN/SNOTEL Data](http://ncss-tech.github.io/AQP/soilDB/fetchSCAN-demo.html)

## Related Packages
 * [aqp](https://github.com/ncss-tech/aqp)
 * [sharpshootR](https://github.com/ncss-tech/sharpshootR)
 

## Examples
```r
library(soilDB)
library(sharpshootR)
library(latticeExtra)

# get morphology + extended summaries
soils <- c('cecil', 'altavista', 'lloyd', 'wickham', 'wilkes',  'chewacla', 'congaree')
s <- fetchOSD(soils, extended = TRUE)

# experimental viz of hillslope position, from SSURGO component records
res <- vizHillslopePosition(s$hillpos)
print(res$fig)

# profile sketches
par(mar=c(0,1,0,4), xpd=NA)
plot(s$SPC, plot.order=res$order, cex.names=1, axis.line.offset = -0.1, width=0.2)


# siblings
s <- 'Amador'
amador <- siblings(s, only.major = FALSE, component.data = TRUE)

# limit to named soil series
sib.data <- subset(amador$sib.data, subset= ! compkind %in% c('Miscellaneous area', 'Family', 'Taxon above family'))

# get parsed OSD records
sibs <- fetchOSD(c(s, unique(amador$sib$sibling)), extended = TRUE)

# order by subgroup taxonomy
# invert colors
par(mar=c(0,0,0,0), fg='white', bg='black')
SoilTaxonomyDendrogram(sibs$SPC, dend.width = 1.5, y.offset = 0.4, scaling.factor = 0.02, width=0.2, cex.taxon.labels = 1, cex.names = 1)
```

## Dependency Graph
![](https://cran.microsoft.com/packagedata/graphs/soilDB.png)



NEWS
-------------------------- soilDB 2.3.9 (2019-07-25) ----------------------------
* SDA_query() no longer writes temporary files, c/o suggestion from Kyle Bocinsky (#106 / #108)
* fetchOSD() gets a sanity check to protect against going over GET request limits
* makeChunks() added to util functions, useful for splitting data before sending API calls

-------------------------- soilDB 2.3.8 (2019-03-22) ----------------------------
* loafercreek, gopheridge, and mineralking sample data have been updated with valid place-holder in @sp
* bug fix for SDA_query() related to multi-line records (https://github.com/ncss-tech/soilDB/issues/28)

-------------------------- soilDB 2.3.7 (2019-03-12) ----------------------------
* sharpshootR added to SUGGESTS
* fetchHenry() and fetchSCAN() now include water year/day (Oct 1 -- Sep 30)
* HenryTimeLine() convenience function added
* bug fix in fetchOSD(..., extended=TRUE) when no climate data available
* bug fix in SDA_query()

-------------------------- soilDB 2.3.6 (2019-02-12) ----------------------------
* simplifyFragmentData() and related functions now 4-5x faster
* fetchOSD() now returns metadata when extended=TRUE

-------------------------- soilDB 2.3.5 (2019-01-14) ----------------------------
* NAMESPACE and R CMD check fixes
* documentation updates
* soilDB now suggests `stringr`
* new tests
* release to CRAN

-------------------------- soilDB 2.3.3 (2018-12-18) ----------------------------
* bug fix in fetchNASIS() related to conversion of NULL fragment volume to 0
* fetchKSSL() can now automatically simplify colors with `simplifyColors = TRUE`

-------------------------- soilDB 2.3 (2018-11-17) ------------------------------
* new function for exploring soil series co-occurrence data: siblings()
* fetchOSD(..., extended=TRUE) gets competing soil series
* new tests
* removed old sample gSSURGO chunk and related documentation
* release to CRAN

-------------------------- soilDB 2.2-6 (2018-10-11) ------------------------------
* new function for spatial queries: SDA_spatialQuery(), still needs testing and documentation

-------------------------- soilDB 2.2-6 (2018-10-11) ------------------------------
* fetchOSD gets an overhaul, new API and features

-------------------------- soilDB 2.2-5 (2018-10-05) ------------------------------
* experimental interface to SoillWeb OSD fulltext search: OSDquery()

-------------------------- soilDB 2.2-4 (2018-09-04) ------------------------------
* numerous bug-fixes in simplifyFragmentData() see (https://github.com/ncss-tech/soilDB/issues/70)

-------------------------- soilDB 2.2-1 (2018-05-21) ------------------------------
* get_mutext_from_NASIS_db() added for extraction of map unit text notes

-------------------------- soilDB 2.2 (2018-05-08) ------------------------------
* CRAN release
* better bug-fix for fetchSCAN and missing data (https://github.com/ncss-tech/soilDB/issues/26)
* exposing some of the internal functionality used by fetchHenry

-------------------------- soilDB 2.1-1 (2018-03-29) ------------------------------
* added more SCAN/SNOTEL metadata (https://github.com/ncss-tech/soilDB/issues/61)

-------------------------- soilDB 2.1 (2018-03-12) ------------------------------
* generalized local NASIS ODBC authentication, should work on windows 7, 8, 10 https://github.com/ncss-tech/soilDB/issues/54

-------------------------- soilDB 2.0-4 (2018-03-06) ----------------------------
* bug fix for subtle change in how SCAN data are returned from webservice

-------------------------- soilDB 2.0-3 (2018-02-13) ----------------------------
* bug fix for simplfyFragmentData(, nullFragsAreZero=FALSE), still more work to do

-------------------------- soilDB 2.0-2 (2018-01-29) ----------------------------
* bug fix for simplfyFragmentData() when fragment volume > 100%

-------------------------- soilDB 2.0-1 (2018-01-23) ----------------------------
* updated `loafercreek` and `gopheridge` sample datasets and manual page to reflect latest fetchNASIS

-------------------------- soilDB 2.0 (2017-12-23) ------------------------------
* bug fixes, manual page updates, and stabilization of NASIS interaction
* released to CRAN

-------------------------- soilDB 1.8.12 (2017-09-19) ----------------------------
* bug fix in fetchHenry(), it is now possible to query only sensor metadata

-------------------------- soilDB 1.8.11 (2017-08-17) ----------------------------
* temporary bug-fix related to SCAN data (https://github.com/ncss-tech/soilDB/issues/26)

-------------------------- soilDB 1.8.10 (2017-08-10) ----------------------------
* re-write of mix_and_clean_colors() so that color mixing happens in CIE LAB space

-------------------------- soilDB 1.8.5 (2017-05-24) ----------------------------
* re-write of SDA_query(), better error handling, support for multi-part result sets

-------------------------- soilDB 1.8.2 (2017-03-24) ----------------------------
* fetchHenry() can now access water level data, see manual page

-------------------------- soilDB 1.8.1 (2017-01-23) ----------------------------
* new function: get_comonth_from_NASIS_db() see manual page for details

-------------------------- soilDB 1.8-10 (2016-12-27) ----------------------------
* converting all URLs to 'HTTPS', addressing bugs with fetchSCAN() and associated functions

-------------------------- soilDB 1.8-7 (2016-11-16) ----------------------------
* sanity checks for fetchHenry()
* SCAN_sensor_metadata() vectorized and documented
* unique row names in SPDF returned by seriesExtent()

-------------------------- soilDB 1.8-6 (2016-11-04) ----------------------------
* converted all links to SoilWeb servers to HTTPS
* converted all NRCS web-service links to HTTPS

-------------------------- soilDB 1.8-4 (2016-10-07) ----------------------------

R/STR.R

#
STRplot <- function(mast, msst, mwst, permafrost=FALSE, pt.cex=2.75, leg.cex=0.85) {
  
  # make a row of rectangles with colors based on STR
  .makeRow <- function(x, y, STR, colors, O, S){
    x.offset <- 0.4
    y.offset <- 0.25
    
    rect(xleft=x - x.offset, 
         xright = x + x.offset, 
         ybottom = rep(y, times=length(x)) - y.offset, 
         ytop = rep(y, times=length(x)) + y.offset, 
         col=colors[as.numeric(STR)])
    
    # annotate conditions
    text(min(x) - 1, y, labels = sprintf('%s | %s', O, S), cex=0.75, font=2, adj=0.5)
  }
  
  
  # reasonable colors for STR
  cols <- c("white", "purple", "blue", "lightblue", "darkgreen", "lightgreen", 
            "orange", "yellow", "brown", "red")
  
  
  
  # can only iterate over a single axis
  if(length(mast) > 1 & length(msst) > 1)
    stop()
  
  # which axis are we iterating over?
  if(length(mast) > 1) {
    
    # iterate over MAST
    x <- mast
    x.label <- 'MAST (C)'
    alt.label <- 'MSST'
    alt.summary <- msst
    
    msst <- rep(msst, times=length(mast))
    mwst <- rep(mwst, times=length(mast))
    
  } else if(length(msst) > 1) {
    
    # iterate over MSST
    x <- msst
    x.label <- 'MSST (C)'
    alt.label <- 'MAST'
    alt.summary <- mast
    
    mast <- rep(mast, times=length(msst))
    mwst <- rep(mwst, times=length(msst))
  }
  
  
  ## estimate STR over 9 possible situations
  # no information
  x.1 <- estimateSTR(mast, msst, mwst, O.hz = rep(NA, times=length(mast)), saturated = rep(NA, times=length(mast)), permafrost=permafrost)
  # O hz information
  x.2 <- estimateSTR(mast, msst, mwst, O.hz = rep(TRUE, times=length(mast)), saturated = rep(NA, times=length(mast)), permafrost=permafrost)
  x.3 <- estimateSTR(mast, msst, mwst, O.hz = rep(FALSE, times=length(mast)), saturated = rep(NA, times=length(mast)), permafrost=permafrost)
  # saturation information
  
  
  #
  x.9 <- estimateSTR(mast, msst, mwst, O.hz = rep(FALSE, times=length(mast)), saturated = rep(FALSE, times=length(mast)), permafrost=permafrost)
  x.8 <- estimateSTR(mast, msst, mwst, O.hz = rep(TRUE, times=length(mast)), saturated = rep(FALSE, times=length(mast)), permafrost=permafrost)
  # 
  x.7 <- estimateSTR(mast, msst, mwst, O.hz = rep(FALSE, times=length(mast)), saturated = rep(TRUE, times=length(mast)), permafrost=permafrost)
  x.6 <- estimateSTR(mast, msst, mwst, O.hz = rep(TRUE, times=length(mast)), saturated = rep(TRUE, times=length(mast)), permafrost=permafrost)
  #
  x.5 <- estimateSTR(mast, msst, mwst, O.hz = rep(NA, times=length(mast)), saturated = rep(FALSE, times=length(mast)), permafrost=permafrost)
  x.4 <- estimateSTR(mast, msst, mwst, O.hz = rep(NA, times=length(mast)), saturated = rep(TRUE, times=length(mast)), permafrost=permafrost)
  
  
  
  # init plot
  plot(x, rep(1, times=length(x.1)), type='n', 
       axes=FALSE, ylab='', xlab='', xlim=c(min(x) - 0.75, max(x) + 0.5), ylim=c(1, 11))
  
  # horizontal and vertical guides
  segments(x0 = x, x1 = x, y0 = 0, y1 = 9.5, lty=3, col='grey')
  segments(y0 = (2:9)-0.4, y1 = (2:9)-0.4, x0 = min(x) - 0.5, x1 = max(x)+0.5, lty=1, col=c('black', 'grey', 'black', 'grey'))
  
  # thematic rectangles with STR information
  .makeRow(x=x, y=1, STR=x.1, colors = cols, O='?', S='?')
  .makeRow(x=x, y=2, STR=x.2, colors = cols, O='O', S='?')
  .makeRow(x=x, y=3, STR=x.3, colors = cols, O='X', S='?')
  .makeRow(x=x, y=4, STR=x.4, colors = cols, O='?', S='S')
  .makeRow(x=x, y=5, STR=x.5, colors = cols, O='?', S='X')
  
  
  .makeRow(x=x, y=6, STR=x.6, colors = cols, O='O', S='S')
  .makeRow(x=x, y=7, STR=x.7, colors = cols, O='X', S='S')
  
  .makeRow(x=x, y=8, STR=x.8, colors = cols, O='O', S='X')
  .makeRow(x=x, y=9, STR=x.9, colors = cols, O='X', S='X')
  
  # label x-axis
  axis(side = 1, at = x, cex.axis=0.75)
  mtext(x.label, side=1, line=2.5, font=2)
  
  # label non-varying parameters
  mtext(sprintf('MWST: %s (C)', unique(mwst)), side=1, at=min(x), line=2.5, adj=0, font=2)
  mtext(sprintf('%s: %s (C)', alt.label, alt.summary), side=1, at=max(x)+0.5, line=2.5, adj=1, font=2)
  
  # labels guide
  mtext('O / Saturation', side=3, at=min(x)+0.25, line=-5, adj=1, font=3, cex=0.75)
  
  # legend
  legend('top', 
         legend=c("gelic", "cryic", "frigid", "isofrigid", "mesic", "isomesic", 
                  "thermic", "isothermic", "hyperthermic", "isohyperthermic"),
         pt.bg=cols,
         pch=22,
         pt.cex=2,
         cex=leg.cex,
         ncol=5,
         inset = 0.05,
         bty='n'
  )
  
  
}



# vectors of MAST, summer mean, winter mean all in Deg C
estimateSTR <- function(mast, mean.summer, mean.winter, O.hz=NA, saturated=NA, permafrost=FALSE) {
  
  # check to make sure that the lengths of vectors are the same
  if(! all.equal(length(mast), length(mean.summer), length(mean.winter)))
    stop('inputs must all have the same length', call. = TRUE)
  
  # iterate over input
  n <- length(mast)
  res <- vector(mode = 'character', length = n)
  
  for(i in seq_along(mast)) {
    # check for NA
    if(any(is.na(c(mast[i], mean.summer[i], mean.winter[i])))){
      res[i] <- NA
      next
    }
    
    # gelic, suborder and GG levels
    if(mast[i] <= 0) {
      res[i] <- 'gelic'
      next
    }
    
    # gelic, order level
    if(mast[i] <= 1 & permafrost) {
      res[i] <- 'gelic'
      next
    }
    
    
    # possibly cryic, because we don't know saturation and O hz status
    if(mast[i] < 8) {
      
      # if we have both saturation and O hz information we can be sure
      if(! is.na(saturated[i]) & ! is.na(O.hz[i])) {
        
        # not saturated
        if(! saturated[i]) {
          
          # no O horizon
          if(! O.hz[i]) {
            if(mean.summer[i] > 0 & mean.summer[i] < 15) {
              res[i] <- 'cryic'
              next
            }
          }
          
          # O horizon
          if(O.hz[i]) {
            if(mean.summer[i] > 0 & mean.summer[i] < 8) {
              res[i] <- 'cryic'
              next
            }
          }
        }
        
        # saturated
        if(saturated[i]) {
          
          # no O horizon
          if(! O.hz[i]) {
            if(mean.summer[i] > 0 & mean.summer[i] < 13) {
              res[i] <- 'cryic'
              next
            }
          }
          
          # O horizon
          if(O.hz[i]) {
            if(mean.summer[i] > 0 & mean.summer[i] < 6) {
              res[i] <- 'cryic'
              next
            }
          }
        }
        
      } 
      
      
      # saturation information only
      # strategy: split the difference O vs. no O horizon
      if(!is.na(saturated[i]) & is.na(O.hz[i])) {
        
        # not saturated
        if(! saturated[i]) {
          if(mean.summer[i] > 0 & mean.summer[i] < 11.5) {
            res[i] <- 'cryic'
            next
          }
        }
        
        # saturated
        if(saturated[i]) {
          if(mean.summer[i] > 0 & mean.summer[i] < 9.5) {
            res[i] <- 'cryic'
            next
          }
        }
        
      }
      
      # O horizon information only
      # strategy: split the difference saturated vs. not saturated
      if(!is.na(O.hz[i]) & is.na(saturated[i])) {
        
        # no O horizon
        if(! O.hz[i]) {
          if(mean.summer[i] > 0 & mean.summer[i] < 14) {
            res[i] <- 'cryic'
            next
          }
        }
        
        # O horizon
        if(O.hz[i]) {
          if(mean.summer[i] > 0 & mean.summer[i] < 7) {
            res[i] <- 'cryic'
            next
          }
        }
        
      }
      
      
      # no additional information
      # assume: saturated during part of summer & no O horizon
      if( is.na(saturated[i] & is.na(O.hz[i]))) {
        if(mean.summer[i] < 13) {
          res[i] <- 'cryic'
          next
        }
      }
     
      
    }
    
    
    ## frigid
    if(mast[i] > 0 & mast[i] < 8) {
      if(mean.summer[i] - mean.winter[i] >= 6) {
        res[i] <- 'frigid'
        next
      }
      else {
        res[i] <- 'isofrigid'
        next
      }
    }
    
    # mesic
    if(mast[i] >= 8 & mast[i] < 15) {
      if(mean.summer[i] - mean.winter[i] >= 6) {
        res[i] <- 'mesic'
        next
      }
      else {
        res[i] <- 'isomesic'
        next
      }
    }
    
    # thermic
    if(mast[i] >= 15 & mast[i] < 22) {
      if(mean.summer[i] - mean.winter[i] >= 6){
        res[i] <- 'thermic'
        next
      }
      else {
        res[i] <- 'isothermic'
        next
      }
    }
    
    # hyperthermic
    if(mast[i] >= 22) {
      if(mean.summer[i] - mean.winter[i] >= 6) {
        res[i] <- 'hyperthermic'
        next
      }
      else {
        res[i] <- 'isohyperthermic'
        next
      }
    }
    
    # unknown
    res[i] <- NA
  }
  
  # set levels
  res <- factor(res, levels=c('gelic', 'cryic','frigid','isofrigid','mesic','isomesic','thermic','isothermic','hyperthermic','isohyperthermic'))
  
  # done
  return(res)
}
